src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs

```diff
@@ -1584,7 +1584,9 @@ private ConsumeNumberResult ConsumeZero(ref ReadOnlySpan<byte> data, ref int i)
             if (nextByte != '.' && nextByte != 'E' && nextByte != 'e')
             {
                 _bytePositionInLine += i;
-                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndOfDigitNotFound, nextByte);
+                ThrowHelper.ThrowJsonReaderException(ref this,
+                    nextByte >= '0' && nextByte <= '9' ? ExceptionResource.InvalidLeadingZeroInNumber : ExceptionResource.ExpectedEndOfDigitNotFound,
```

Comment:
nit: consider using `IsInRangeInclusive` helper to reduce the number of branches/code size of the method.

[sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeAdAEoCuAdjAJZhzUDCA9mAA4sANogDKiAG4sAxnCgBuYsSwBmSjnKdyAb2Lk9lVVgAs5ALJ4AFCL4RGAHmABPeAD5yAEwgwISci2Z+vv4w5AD6TvAACtxQLKzcjACSjAAy/nAAlNq6+rkRcOSMcAAeMABCzgUAvB5eEADaLAC6CkS5uSwAZuQWRaUV8OQAhDUA5NSj5ABkU4Ul5ZXDYwCikzNz/Ysj5KNwoxk57dltR+3hldGx8Ump6eQA1DUsraftACoAFgjcAO4AUlAErQ4BB3IhlsUZHxrr15gMCi4xqQ1rM+gtBnYxgBOSYAfnIEKh12BgPoCBk1GS4ggghY7hSIPc/gA5gAtRDcZIAOXoYGAiHIIAJkLg0JYQNk3DJFIhfDgUng7mWjHcAHlOgARFjMuJc7gwABiUpVSEOr3aaPhGRepwAvmb7Sc9GaVJQTKYcFYbPZ8m5PN4ggEWIGQucojE4uKbmkilkdE72vkNujqrVvI0WmaOt1YZtBttxijk/CljtVtNUXCtmM9gcE7l4+b9GG4JdIwlkjGCo8/Dam59vv9AYxgaDwSKxQkLIkoMlaLZmXBklJBPRYuI4LmU75RsidzisvjCaLiZLpRwqTS6QzQSz2d9ubz+QhBcKiVGSVLyRxZfLFcq1U1bVdX1I0mHcU16ybYtKmtLN9EdXJEP0F1DHdZQvVsBxKj9OoQ0CPwAhbNtrk7dI43gvQk0tRYan9Bpmj7douh6Gj8zGCYKxg9iyyLNiCgLWtKOOaCwnyEiozIooHieJjXgHX4AQlMcEGPSdGAsNTT1Jb9qF/BU4CVFV1S1HUYD1Q1jQg4TXn4uCoL0ZDHLNFyE3qUw4BgD5uHcRJ+EECwPK8ny/L4QRVXUqBqD1ZJaUYFkMiaVC3XIBSh2UsFVInGEtI/M9v3IBBZF8aiq3gOsG2ErzBzmH43xPKMLAAEgAIi0IqoFtJAtH420WvspDXNydzPO83z/MC0aQv8iLriigBBZlmQ69cl0YOKEqShNXSoAA2chgG4bhBHIGc5wXNaVzXFgNwsehgnIalVzgXx7oCQRfkQMorNeh76D4OUEG+8CsiqNwLAsJ76AKABacgPp+L6rKyTEen+wHgZVcg4YRpGQfsx0lFUOBGF5er1M/c9iHjS9aXpRk7w5R8+UQSD9P/YygLMiywJNYhbSAA==)

```suggestion
                    JsonHelpers.IsInRangeInclusive(nextByte, '0', '9') ? ExceptionResource.InvalidLeadingZeroInNumber : ExceptionResource.ExpectedEndOfDigitNotFound,
```